remote-query/remote-query-server/src/main/java/org/infinispan/query/remote/impl/ProtobufMetadataManagerImpl.java

```diff
@@ -55,32 +56,40 @@
 
    private ObjectName objectName;
 
-   private final SerializationContext serCtx;
+   private SerializationContext serCtx;
 
    private EmbeddedCacheManager cacheManager;
 
-   public ProtobufMetadataManagerImpl() {
-      Configuration.Builder cfgBuilder = Configuration.builder();
-      IndexingMetadata.configure(cfgBuilder);
-      serCtx = ProtobufUtil.newSerializationContext(cfgBuilder.build());
-      try {
-         MarshallerRegistration.registerMarshallers(serCtx);
-      } catch (IOException | DescriptorParserException e) {
-         throw new CacheException("Failed to initialise the Protobuf serialization context", e);
-      }
-   }
+   private PersistenceMarshaller persistenceMarshaller;
 
    /**
     * Defines the configuration of the ___protobuf_metadata internal cache.
     */
    @Inject
-   protected void init(EmbeddedCacheManager cacheManager, InternalCacheRegistry internalCacheRegistry) {
+   protected void init(EmbeddedCacheManager cacheManager, InternalCacheRegistry internalCacheRegistry, PersistenceMarshaller pm) {
       this.cacheManager = cacheManager;
+      this.persistenceMarshaller = pm;
+
+      Configuration.Builder cfgBuilder = Configuration.builder();
+      IndexingMetadata.configure(cfgBuilder);
+      persistenceMarshaller.setProtobufConfiguration(cfgBuilder.build());
+
       internalCacheRegistry.registerInternalCache(PROTOBUF_METADATA_CACHE_NAME,
             getProtobufMetadataCacheConfig().build(),
             EnumSet.of(Flag.USER, Flag.PROTECTED, Flag.PERSISTENT));
    }
 
+   @Start
+   public void start() {
+      this.serCtx = persistenceMarshaller.getSerializationContext();
```

Comment:
The ProtobufMetadataManager is used to manage user provided protobuf files (backed by the internal __protobuf_metadata cache), I don't think it should reuse the same serCtx as the persistence marshaller.
Also, it seems weird having the persistence marshaller in the remote query server :)

Reply:
The reason that the serialization contexts is shared was to allow `ProtostreamSerializationContextInitializer` implementations to register protofiles with both the query and persistence context simultaneously. 

Originally I had created a implementation of the SerializationContext interface to wraps the two contexts and ensures that any registering/unregistering of marshallers happens on both contexts. However, the problem is that the protostream implementation casts to `SerializationContextImpl` in a few places, so this approach did not work. 

I agree that it's strange that the persistence marshaller is in the remote query server :see_no_evil:. So I guess the question is how best to abstract the registration of protobuf files with multiple contexts? I could create `InfinispanSerializationContext` which delegates to the two `SerializationContext` instances and expose this via `ProtostreamSerializationContextInitializer`? We would have to deprecate the previous `init` method though.

Reply:
My understanding was that ```ProtostreamSerializationContextInitializer``` is a user facing construct to allow registering extra user marshallers, and the persistence serCtx is a private context that works only with Infinispan internal classes and treats user data as opaque byte[]. This private content is known at "boot" time, and no extra proto/classes are needed at runtime (thus no registration). 

Is there a use case of exposing the persistence ctx to the user, so that it can add, remove and change marshallers/protofiles?

Reply:
> Is there a use case of exposing the persistence ctx to the user, so that it can add, remove and change marshallers/protofiles?

Good point and I don't think there is. I was getting ahead of myself/confused by the idea of having a protobuf based user marshaller, but that's for embedded mode anyway.